onnxruntime/core/providers/cpu/tensor/transpose.cc

```diff
@@ -25,72 +25,140 @@ size_t ComputeOffset(const std::vector<int64_t>& index, const std::vector<size_t
   for (int64_t k = num_axes - 1; k >= 0; --k) {
     index[k]++;
     if (index[k] < upper_bound[k]) break;
     index[k] = 0;
   }
 }
 
+// DoTransposeSingleBlock: specialization of DoTranspose for the num_blocks=1 case.
+// copies source tensor to target, transposing elements.
+static inline void DoTransposeSingleBlock(size_t num_elts_in_block, const void* source, void* target,
+                                          size_t element_size) {
+  size_t blocksize = num_elts_in_block * element_size;
+  // copy
+  memcpy(target, source, blocksize);
+}
+
+static inline void DoTransposeSingleBlock(size_t num_elts_in_block, const std::string* source, std::string* target) {
+  const std::string* end = source + num_elts_in_block;
+  std::copy(source, end, target);
+}
+
 // DoTranspose: copies source tensor to target, transposing elements.
 // The stride vector indicates the transposition.
-template <typename T>
 static void DoTransposeImpl(int64_t num_axes, const std::vector<int64_t>& target_dims,
                             size_t num_blocks, size_t num_elts_in_block, const std::vector<size_t>& stride,
-                            const T* source, T* target) {
-  size_t blocksize = num_elts_in_block * sizeof(float);
+                            const uint8_t* source, uint8_t* target, size_t element_size) {
+  size_t blocksize = num_elts_in_block * element_size;
+  // index used to iterate over target iteration-space
+  std::vector<int64_t> target_index(num_axes, 0);
+  for (size_t i = 0; i < num_blocks; ++i) {
+    // convert target_index into an offset in source data
+    size_t source_offset = ComputeOffset(target_index, stride, num_axes);
+
+    // copy
+    memcpy(target, source + source_offset * element_size, blocksize);
+
+    // increment target_index:
+    IncrementIndex(target_index, target_dims, num_axes);
+    target += blocksize;
+  }
+}
+
+static void DoTransposeImpl(int64_t num_axes, const std::vector<int64_t>& target_dims,
+                            size_t num_blocks, size_t num_elts_in_block, const std::vector<size_t>& stride,
+                            const std::string* source, std::string* target) {
   // index used to iterate over target iteration-space
   std::vector<int64_t> target_index(num_axes, 0);
   for (size_t i = 0; i < num_blocks; ++i) {
     // convert target_index into an offset in source data
     size_t source_offset = ComputeOffset(target_index, stride, num_axes);
 
     // copy
-    memcpy(target, source + source_offset, blocksize);
+    DoTransposeSingleBlock(num_elts_in_block, source + source_offset, target);
 
     // increment target_index:
     IncrementIndex(target_index, target_dims, num_axes);
     target += num_elts_in_block;
   }
 }
 
+template <class T>
+inline void CopyPrim(const uint8_t* source, uint8_t* target) {
+  *reinterpret_cast<T*>(target) = *reinterpret_cast<const T*>(source);
+}
+
+static inline void CopyPrimitiveElement(const uint8_t* source, uint8_t* target, size_t element_size) {
```

Comment:
Personally, I don't think this implementation is good. You do the switch...case for each element, which is totally unnecessary and only need be done once.

Reply:
Good point

---
In reply to: [264978950](https://github.com/Microsoft/onnxruntime/pull/561#discussion_r264978950) [](ancestors = 264978950)